bson-codec-design/DESIGN.md

```diff
@@ -0,0 +1,157 @@
+<dt>Proposed Design</dt>
+<dd>The design laid out in this document</dd>
+</dl>
+
+## Motivation
+The v1 design for codec processing in the BSON library made a number of assumptions about the
+control the user had over their types, the amount of granular control users would want to have over
+codec processing, and the flexibility required during codec processing. A key initial assumption of
+the v1 design was that users who needed to handle types in a custom way would be able to make those
+types implement the `bson.Marshaler` and `bson.Unmarshaler` interfaces. This turned out to be too
+restrictive because users sometimes want to handle the processing for types they do not define
+themselves. For example, users who generate types using a protocol buffer library
+or users who want to handle the processing of types in the standard library.
+
+The design of the v1 encoder and decoder for the Go driver presents problems when users wish to
+have more control over how specific types are handled when encoding or decoding. The control that
+users require exists on two main levels:
+- control over how individual types are encoded or decoded
+- control over how tags are processed
+
+The granularity of control users require can be defined in three levels:
+- Control over how individual types are encoded and decoded when they are embedded in another type,
+such as a map or a struct
+- Control over how individual types are encoded and decoded when they are provided directly to the
+encoder or decoder
+- Control over how tags are processed and the parameters that are returned
+
+The proposed design allows for the first two granularities to be handled together.
+
+## High Level Design
+### Registries
+To enable users to control how types are encoded and decoded, the proposed design uses a type
+registry. This registry will be provided in any context where encoding or decoding occurs. The registry
+will be used to lookup a type's codec, which will then be used for the actual encoding a decoding.
+This will allow users to control nearly all aspects of the encoding and decoding process. One
+failing of the v1 design is that users cannot control the encoding and decoding of types
+they do not own, e.g. `time.Time`. The proposed design handles this by allowing users to register a
+codec for any type.
+
+Pointer and value types are handled separately in the registry as this simplifies the registry and
+the implementations of codecs, since they will only have to worry about a single type and not the
+type plus a pointer to that type. While likely not useful, there is no design
+reason to prevent users from registering codecs for internal types like `bson.Element`,
+`bson.Document`, or `bson.Reader`. Users can override the functionality of built in types as
+well. The registry will be used for all encoding and decoding operations, since the cost of calling
+`reflect.TypeOf` is inexpensive.
```

Comment:
What type is this registry? Is this a type or an interface? Will we provide a default registry and allow people to append to it?

Reply:
It should be a type. Someone can't pass us an interface and ask us to marshal it. It has an underlying type and that type is what get's looked up. If there is no type registered, we'll look through the interface registry and see if the type implements any of the registered interfaces; if so, we'll invoke the hook associated with it. Note that the interface scanning is likely much more expensive and it may be pertinent to cache the results per lookup type, but that's an implementation detail for later.

Reply:
I was asking more about the registry itself not the thing we were marshalling. 

It's reasonably common to pass interface types to the marshaller (the underlying types should be marshalale, and are with other libraries), and if you have a way of constructing a destination object of the correct type that is passed as an interface type, the round trip can/should work (evergreen uses this behavior.)

Reply:
It's a struct, not an interface